### DISCUSSION TITLE
Feature Request: Sparse Vector/List DType

### DISCUSSION THREAD

**Discussion by ghodouss** [duckdb/duckdb#14822] [Ideas]
Hi. Absolutely love the duckdb library! I started with it locally and now use it extensively in multiple production systems.

I have a set of models that output sparse high dimensional vectors (dropping indicies with 0). To save space, I store these in a sparse format and have a series of macros for working with them.

```
CREATE OR REPLACE TEMPORARY MACRO sparse_list_extract(position, inds, vals) AS  COALESCE(vals[list_position(inds, position)], 0);
CREATE OR REPLACE TEMPORARY MACRO sparse_list_select(positions, inds, vals) AS  list_transform(positions, x-> sparse_list_extract(x, inds, vals));
CREATE OR REPLACE TEMPORARY MACRO dense_x_sparse_dot_product(arr, inds, vals) AS list_dot_product(list_select(arr, inds)::DOUBLE[], vals::DOUBLE[]);
CREATE OR REPLACE TEMPORARY MACRO sparse_to_dense(K, inds, vals) AS list_transform(
        generate_series(1, K), x-> coalesce(list_extract(vals, list_position(inds, x)), 0))
```

This is pretty painless, but I could imagine it being a really nice feature a sparse data type was implemented under the hood and the client could just treat the vectors as dense and the sparsity was implemented under the hood. Scipy supports this by storing vectors as a tuple of (dim: int, inds: list[int], values: list[float]).

**Comment by mkmcc** [comment]
This would be a phenomenal feature to add to duckdb.  I work in data science, and duckdb is so performant that disk space on the server is the bottleneck for my project.  Sparse arrays aren't needed for every project, but when they are appropriate they can reduce the storage by very large multiples.

I tried something analogous to this solution, using scipy sparse arrays and a python udf, but it was very slow compared to native duckdb.
